### ISSUE TITLE
App raises fatal error about missing model on recent activations

### ISSUE THREAD

**Issue by billyonthemountain** [nextcloud/suspicious_login#82]
On new installs, the app will raises a fatal error about it missing a model:
```
{"reqId":"ksqdCUucdhnLY2UHkYM1","level":4,"time":"2019-05-17T15:16:15+00:00","remoteAddr":"10.8.10.6","user":"admin","app":"suspicious_login","method":"POST","url":"\/index.php\/login?redirect_url=\/index.php\/settings\/admin\/logging","message":"No models found. Can't load latest","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:66.0) Gecko\/20100101 Firefox\/66.0","version":"16.0.1.1"}
{"reqId":"ksqdCUucdhnLY2UHkYM1","level":2,"time":"2019-05-17T15:16:15+00:00","remoteAddr":"10.8.10.6","user":"admin","app":"suspicious_login","method":"POST","url":"\/index.php\/login?redirect_url=\/index.php\/settings\/admin\/logging","message":"Could not predict suspiciousness: No models found","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:66.0) Gecko\/20100101 Firefox\/66.0","version":"16.0.1.1"}
```
This is obviously expected since there has not been enough time to amass data. Would it not make sense to adapt the severity of the message based on the presence of "young" training data ? Eg. something like:

1. detect the presence of data and its age
2. if no data or data is < 60 days old
 => warn
3. else (if no model and data is >= 60 days old) 
 => assume that a model should have been trained and raise a (fatal) error.

This would be consistent when the "severity" of the message if the admin console:
> No classifier model has been trained yet. This most likely means that you just enabled the app recently. Because the training of a model requires good data, the app waits until logins of at least 60 days have been captured.

**Comment by ChristophWurst** [comment]
> On new installs, the app will raises a fatal error about it missing a model:

When?

**Comment by ChristophWurst** [comment]
Did you run the `predict` occ command? The login classification logic would actually catch that error.

**Comment by billyonthemountain** [comment]
> > On new installs, the app will raises a fatal error about it missing a model:
> 
> When?

Generally at every login and otherwise also when requesting the calendars, I get the the fatal error and warning messages in the logs. Here are two examples:
```
{"reqId":"ksqdCUucdhnLY2UHkYM1","level":4,"time":"2019-05-17T15:16:15+00:00","remoteAddr":"10.8.10.6","user":"admin","app":"suspicious_login","method":"POST","url":"\/index.php\/login?redirect_url=\/index.php\/settings\/admin\/logging","message":"No models found. Can't load latest","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:66.0) Gecko\/20100101 Firefox\/66.0","version":"16.0.1.1"}
{"reqId":"ksqdCUucdhnLY2UHkYM1","level":2,"time":"2019-05-17T15:16:15+00:00","remoteAddr":"10.8.10.6","user":"admin","app":"suspicious_login","method":"POST","url":"\/index.php\/login?redirect_url=\/index.php\/settings\/admin\/logging","message":"Could not predict suspiciousness: No models found","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:66.0) Gecko\/20100101 Firefox\/66.0","version":"16.0.1.1"}
{"reqId":"4kGasg3p4vTIcv8BuCi5","level":4,"time":"2019-05-17T21:19:22+00:00","remoteAddr":"86.62.85.51","user":"<USER_NAME>","app":"suspicious_login","method":"PROPFIND","url":"\/remote.php\/dav\/calendars\/<USER_NAME>\/contact_birthdays\/","message":"No models found. Can't load latest","userAgent":"DAVx5\/2.2.3.1-ose (2019\/01\/27; dav4jvm; okhttp\/3.12.1) Android\/8.0.0","version":"16.0.1.1"}
{"reqId":"4kGasg3p4vTIcv8BuCi5","level":2,"time":"2019-05-17T21:19:22+00:00","remoteAddr":"86.62.85.51","user":"<USER_NAME>","app":"suspicious_login","method":"PROPFIND","url":"\/remote.php\/dav\/calendars\/<USER_NAME>\/contact_birthdays\/","message":"Could not predict suspiciousness: No models found","userAgent":"DAVx5\/2.2.3.1-ose (2019\/01\/27; dav4jvm; okhttp\/3.12.1) Android\/8.0.0","version":"16.0.1.1"}
```

> Did you run the `predict` occ command? The login classification logic would actually catch that error.

Nope, I just did a "vanilla" install without doing anything else. Running `occ suspiciouslogin:predict <MYUSER> 10.8.10.6` I get:

```
In ModelPersistenceService.php line 91:
                   
  No models found  
                   
In QBMapper.php line 248:
                                                                                                     
  Did expect one result but found none when executing: query "SELECT * FROM `*PREFIX*suspicious_log  
  in_model` ORDER BY `created_at` desc LIMIT 1";                                                     
                                                                                                     
suspiciouslogin:predict <uid> <ip> [<model>]
```
and in the logs
```
{"reqId":"87cw1dSLmql3nT5IyJ7O","level":4,"time":"2019-05-20T11:51:42+00:00","remoteAddr":"","user":"--","app":"suspicious_login","method":"","url":"--","message":"No models found. Can't load latest","userAgent":"--","version":"16.0.1.1"}
```

**Comment by ChristophWurst** [comment]
I think I got it. The estimator is invoked here: https://github.com/nextcloud/suspicious_login/blob/2ce88ec0652fb5b15430eae9ca9b0eaa7f20fac2/lib/Service/LoginClassifier.php#L111. Which thrown a `ServiceException` when no model can be loaded. This is caught and logged here: https://github.com/nextcloud/suspicious_login/blob/2ce88ec0652fb5b15430eae9ca9b0eaa7f20fac2/lib/Service/LoginClassifier.php#L115-L119.

This is not a fatal error, but a warning. Can you confirm? :)

**Comment by ChristophWurst** [comment]
Aaaand there is a second log statement at https://github.com/nextcloud/suspicious_login/blob/2ce88ec0652fb5b15430eae9ca9b0eaa7f20fac2/lib/Service/ModelPersistenceService.php#L90. The latter should probably be of level debug. Mind sending a pull request to fix this? All you need to do is change the method name :)

**Comment by ChristophWurst** [comment]
Fixed with #87
